File: 2024/day_17/solution.py
def detect_bombs(grid: list[list[bool]]) -> list[list[int]]:
  directions = [
    (-1, -1), (-1, 0), (-1, 1),
    (0, -1),          (0, 1),
    (1, -1), (1, 0), (1, 1)
  ]

  result = []

  for row_index, row in enumerate(grid):
    row_result = []
    
    for column_index in range(len(row)):
      count = 0
      
      for dx, dy in directions:
        x, y = row_index + dx, column_index + dy

        if 0 <= x < len(grid) and 0 <= y < len(row) and grid[x][y]:
          count += 1
      
      row_result.append(count)

    result.append(row_result)

  return result

File: 2024/day_17/solution.ts
(() => {
  function detectBombs(grid: boolean[][]): number[][] {
    const directions: [number, number][] = [
      [-1, -1],
      [-1, 0],
      [-1, 1],
      [0, -1],
      [0, 1],
      [1, -1],
      [1, 0],
      [1, 1],
    ];

    const result: number[][] = [];

    grid.forEach((row, rowIndex) => {
      const rowResult = row.map((_, columnIndex) => {
        return directions.reduce((count, [dx, dy]) => {
          const x = rowIndex + dx;
          const y = columnIndex + dy;
          return count + (grid[x] && grid[x][y] ? 1 : 0);
        }, 0);
      });

      result.push(rowResult);
    });

    return result;
  }

  console.log(
    detectBombs([
      [true, false, false],
      [false, true, false],
      [false, false, false],
    ])
  );
  // [
  //   [1, 2, 1],
  //   [2, 1, 1],
  //   [1, 1, 1]
  // ]

  console.log(
    detectBombs([
      [true, false],
      [false, false],
    ])
  );
  // [
  //   [0, 1],
  //   [1, 1]
  // ]

  console.log(
    detectBombs([
      [true, true],
      [false, false],
      [true, true],
    ])
  );
  // [
  //   [1, 1],
  //   [4, 4],
  //   [1, 1]
  // ]
})();
